official/recommendation/ncf_keras_main.py

```diff
@@ -0,0 +1,241 @@
+    if FLAGS.tpu:
+      raise ValueError("NCF in Keras does not support TPU for now")
+    run_ncf(FLAGS)
+
+
+def run_ncf(_):
+  """Run NCF training and eval with Keras."""
+  params = ncf_common.parse_flags(FLAGS)
+
+  distribution = ncf_common.get_distribution_strategy(params)
+
+  num_users, num_items, num_train_steps, num_eval_steps, producer = (
+      ncf_common.get_inputs(params))
+
+  params["num_users"], params["num_items"] = num_users, num_items
+  producer.start()
+  model_helpers.apply_clean(flags.FLAGS)
+
+  # TODO(shiningsun): Both MirroredStrategy and OneDeviceStrategy error out.
+  # Find out why and change the distribute to distribution
+  with distribution_utils.MaybeDistributionScope(None):
+    keras_model = _get_keras_model(params)
+
+    optimizer = ncf_common.get_optimizer(params)
+    keras_model.compile(optimizer=optimizer)
+
+    train_input_dataset, eval_input_dataset = _get_train_and_eval_data(producer, params)
+
+    keras_model.fit(train_input_dataset,
+        epochs=FLAGS.train_epochs,
+        steps_per_epoch=num_train_steps,
+        callbacks=[IncrementEpochCallback(producer)],
+        verbose=2)
+
+    tf.logging.info("Training done. Start evaluating")
+
+    eval_results = keras_model.evaluate(
+        eval_input_dataset,
+        steps=num_eval_steps,
+        verbose=2)
+
+    tf.logging.info("Keras evaluation is done.")
+
+  return eval_results
+
+
+def _get_keras_model(params):
+  batch_size = params['batch_size']
+  user_input = tf.keras.layers.Input(
+      shape=(), batch_size=batch_size, name=movielens.USER_COLUMN, dtype=tf.int32)
+  item_input = tf.keras.layers.Input(
+      shape=(), batch_size=batch_size, name=movielens.ITEM_COLUMN, dtype=tf.int32)
+
+  # Dummy duplicate mask. We need it because it is part of the eval input
+  dup_mask_input = tf.keras.layers.Input(
+      shape=(), batch_size=batch_size, name=rconst.DUPLICATE_MASK, dtype=tf.int32)
+  # Labels as input for the custom loss function
+  labels_input = tf.keras.layers.Input(
+      shape=(), batch_size=batch_size, name="labels", dtype=tf.int32)
+  # valid_point_mask as input for the custom loss function
+  valid_pt_mask_input = tf.keras.layers.Input(
+      shape=(), batch_size=batch_size, name=rconst.VALID_POINT_MASK, dtype=tf.int32)
+
+  base_model = neumf_model.construct_model(user_input, item_input, params)
+
+  keras_model_inputs = base_model.inputs
+
+  keras_model_inputs.append(dup_mask_input)
+  keras_model_inputs.append(labels_input)
+  keras_model_inputs.append(valid_pt_mask_input)
+
+  keras_model = tf.keras.Model(
+      inputs=keras_model_inputs,
+      outputs=base_model.outputs)
+  keras_model.summary()
+
+  logits = keras_model.output
+  softmax_logits = ncf_common.convert_to_softmax_logits(logits)
+
+  _add_custom_loss(
+      keras_model,
+      logits,
+      softmax_logits,
+      labels_input,
+      tf.cast(valid_pt_mask_input, tf.float32))
+
+  _add_custom_metric(
```

Comment:
Also this would be more readable if structured as

metric = _get_custom_metric(logits, mask, ...)
model.add_metric(metric)

Reply:
Done.